src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,157 @@
+
+    private static class Node<T> {
+        private T value;
+        private Node<T> prev;
+        private Node<T> next;
+
+        Node(T value, Node<T> prev, Node<T> next) {
+            this.value = value;
+            this.prev = prev;
+            this.next = next;
+        }
+    }
+
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
     @Override
     public void add(T value) {
+        Node<T> nodeToAdd = new Node<T>(value, tail, null);
+        if (tail == null) {
+            head = nodeToAdd;
+        } else {
+            tail.next = nodeToAdd;
+        }
+        tail = nodeToAdd;
+        size++;
     }
 
     @Override
     public void add(T value, int index) {
+        checkIndex(index, true);
+        if (index == size) {
+            add(value);
+        } else {
+            Node<T> current = getNode(index);
+            Node<T> newNode = new Node<>(value, current.prev, current);
+
+            if (current.prev == null) {
+                head = newNode;
+            } else {
+                current.prev.next = newNode;
+            }
+            current.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public void addAll(List<T> list) {
+        for (T value: list) {
+            add(value);
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        checkIndex(index, false);
+        return getNode(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        checkIndex(index, false);
+        Node<T> current = getNode(index);
+        T oldValue = current.value;
+        current.value = value;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        checkIndex(index, false);
+        Node<T> current = getNode(index);
+        T removedValue = current.value;
+        unlink(current);
+        return removedValue;
+
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = head;
+
+        while (current != null) {
+            if (object == null ? current.value == null : object.equals(current.value)) {
+                unlink(current);
+                return true;
+            }
+            current = current.next;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size == 0;
+    }
+
+    private void checkIndex(int index, boolean isForAdd) {
+        if (isForAdd) {
+            if (index < 0 || index > size) {
+                throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+            }
+        } else {
+            if (index < 0 || index >= size) {
+                throw new IndexOutOfBoundsException("Index: " + index + ", Size: " + size);
+            }
+        }
```

Comment:
Consider creating a separate method 'checkIndexForAdd' and 'checkIndex' to avoid the boolean parameter 'isForAdd'. This will make the code cleaner and separate the logic for checking indices when adding and when accessing elements.